File: backend/app/models/signal.py
from sqlalchemy import Column, Integer, String, Float, DateTime, Boolean, Enum, JSON, Text
from sqlalchemy.sql import func
import enum
from app.database import Base


class SignalStatus(str, enum.Enum):
    ACTIVE = "ACTIVE"
    EXECUTED = "EXECUTED"
    EXPIRED = "EXPIRED"
    REJECTED = "REJECTED"


class Signal(Base):
    __tablename__ = "signals"

    id = Column(Integer, primary_key=True, index=True)
    symbol = Column(String(20), nullable=False, index=True)
    market = Column(String(10), nullable=False)
    timeframe = Column(String(10), nullable=False)
    direction = Column(String(10), nullable=False)
    status = Column(Enum(SignalStatus), default=SignalStatus.ACTIVE, index=True)

    # Price levels
    entry_zone_low = Column(Float)
    entry_zone_high = Column(Float)
    stop_loss = Column(Float)
    tp1 = Column(Float)
    tp2 = Column(Float)
    tp3 = Column(Float)

    # Strategy components
    liquidity_sweep_detected = Column(Boolean, default=False)
    bos_detected = Column(Boolean, default=False)
    pullback_confirmed = Column(Boolean, default=False)
    sweep_level = Column(Float)
    bos_level = Column(Float)

    # ATR
    atr_value = Column(Float)

    # AI scoring
    confidence_score = Column(Float)
    feature_importance = Column(JSON)
    model_version = Column(String(20))

    # Context
    market_structure = Column(String(50))
    session = Column(String(20))
    notes = Column(Text)

    created_at = Column(DateTime(timezone=True), server_default=func.now())
    expires_at = Column(DateTime(timezone=True))
    executed_at = Column(DateTime(timezone=True))

File: frontend/lib/models/signal.dart
class Signal {
  final int id;
  final String symbol;
  final String market;
  final String timeframe;
  final String direction;
  final String status;
  final double? entryZoneLow;
  final double? entryZoneHigh;
  final double? stopLoss;
  final double? tp1;
  final double? tp2;
  final double? tp3;
  final bool liquiditySweepDetected;
  final bool bosDetected;
  final bool pullbackConfirmed;
  final double? sweepLevel;
  final double? bosLevel;
  final double? atrValue;
  final double? confidenceScore;
  final Map<String, dynamic>? featureImportance;
  final String? marketStructure;
  final String? session;
  final DateTime? createdAt;

  const Signal({
    required this.id,
    required this.symbol,
    required this.market,
    required this.timeframe,
    required this.direction,
    required this.status,
    this.entryZoneLow,
    this.entryZoneHigh,
    this.stopLoss,
    this.tp1,
    this.tp2,
    this.tp3,
    required this.liquiditySweepDetected,
    required this.bosDetected,
    required this.pullbackConfirmed,
    this.sweepLevel,
    this.bosLevel,
    this.atrValue,
    this.confidenceScore,
    this.featureImportance,
    this.marketStructure,
    this.session,
    this.createdAt,
  });

  factory Signal.fromJson(Map<String, dynamic> json) {
    return Signal(
      id: json['id'] as int,
      symbol: json['symbol'] as String,
      market: json['market'] as String,
      timeframe: json['timeframe'] as String,
      direction: json['direction'] as String,
      status: json['status'] as String,
      entryZoneLow: json['entry_zone_low'] != null ? (json['entry_zone_low'] as num).toDouble() : null,
      entryZoneHigh: json['entry_zone_high'] != null ? (json['entry_zone_high'] as num).toDouble() : null,
      stopLoss: json['stop_loss'] != null ? (json['stop_loss'] as num).toDouble() : null,
      tp1: json['tp1'] != null ? (json['tp1'] as num).toDouble() : null,
      tp2: json['tp2'] != null ? (json['tp2'] as num).toDouble() : null,
      tp3: json['tp3'] != null ? (json['tp3'] as num).toDouble() : null,
      liquiditySweepDetected: json['liquidity_sweep_detected'] as bool? ?? false,
      bosDetected: json['bos_detected'] as bool? ?? false,
      pullbackConfirmed: json['pullback_confirmed'] as bool? ?? false,
      sweepLevel: json['sweep_level'] != null ? (json['sweep_level'] as num).toDouble() : null,
      bosLevel: json['bos_level'] != null ? (json['bos_level'] as num).toDouble() : null,
      atrValue: json['atr_value'] != null ? (json['atr_value'] as num).toDouble() : null,
      confidenceScore: json['confidence_score'] != null ? (json['confidence_score'] as num).toDouble() : null,
      featureImportance: json['feature_importance'] as Map<String, dynamic>?,
      marketStructure: json['market_structure'] as String?,
      session: json['session'] as String?,
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at'] as String) : null,
    );
  }

  bool get isLong => direction == 'LONG';
  bool get isActive => status == 'ACTIVE';
  bool get isHighConfidence => (confidenceScore ?? 0) >= 0.75;
  int get confluenceCount =>
      (liquiditySweepDetected ? 1 : 0) +
      (bosDetected ? 1 : 0) +
      (pullbackConfirmed ? 1 : 0);
}
